cocos/renderer/gfx-gles3/GLES3Context.cpp

```diff
@@ -137,18 +137,106 @@ bool GLES3Context::initialize(const ContextInfo &info) {
         _colorFmt = Format::RGBA8;
         _depthStencilFmt = Format::D24S8;
 
-        const EGLint attribs[] = {
+        bool msaaEnabled = _device->hasFeature(Feature::MSAA);
+        EGLint redSize{8}, greenSize{8}, blueSize{8}, alphaSize{8}, depthSize{24}, stencilSize{8}, sampleBufferSize{msaaEnabled ? EGL_DONT_CARE : 0}, sampleSize{msaaEnabled ? EGL_DONT_CARE : 0};
+
+        // TODO: there are more options if not limit rgba to 8.
+        EGLint defaultAttribs[] = {
             EGL_SURFACE_TYPE, EGL_WINDOW_BIT | EGL_PBUFFER_BIT,
             EGL_RENDERABLE_TYPE, EGL_OPENGL_ES3_BIT_KHR,
             // EGL_BUFFER_SIZE, colorBuffSize,
-            EGL_BLUE_SIZE, 8,
-            EGL_GREEN_SIZE, 8,
-            EGL_RED_SIZE, 8,
-            EGL_ALPHA_SIZE, 8,
-            EGL_DEPTH_SIZE, 24,
-            EGL_STENCIL_SIZE, 8,
+            EGL_BLUE_SIZE, blueSize,
+            EGL_GREEN_SIZE, greenSize,
+            EGL_RED_SIZE, redSize,
+            EGL_ALPHA_SIZE, alphaSize,
+            EGL_DEPTH_SIZE, depthSize,
+            EGL_STENCIL_SIZE, stencilSize,
+            EGL_SAMPLE_BUFFERS, sampleBufferSize,
+            EGL_SAMPLES, sampleSize,
             EGL_NONE};
 
+        int numConfig = 0;
+        if (eglChooseConfig(_eglDisplay, defaultAttribs, NULL, 0, &numConfig)) {
+            _eglConfigList = new EGLConfig[numConfig];
+        } else {
+            CC_LOG_ERROR("Query configuration failed.");
+            return false;
+        }
+
+        int count = numConfig;
+        if (eglChooseConfig(_eglDisplay, defaultAttribs, _eglConfigList, count, &numConfig) == EGL_FALSE || !numConfig) {
+            CC_LOG_ERROR("eglChooseConfig configuration failed.");
+            return false;
+        }
+
+        enum Performance {
+            HIGH_QUALITY,
+            LOW_POWER,
+            //BALANCE,
+        };
+
+        EGLint depth{0}, stencil{0}, sampleBuffers{0}, sampleCount{0};
+
+        // user defined, how muny attrs you want to filter by yourself.
+        const uint8_t attrNums = 8;
+        int64_t lastScore{0};
+        int params[attrNums] = {0};
+
+        const Performance performanceStrategy = Performance::LOW_POWER;
+        const bool performancePreferred = performanceStrategy == Performance::HIGH_QUALITY;
+        /* auto compare = performanceStrategy ==  Performance::HIGH_QUALITY ? (>) : (<) ;*/
+        // egl configurations store in "ascending order": https://www.khronos.org/registry/EGL/sdk/docs/man/html/eglChooseConfig.xhtml
+        // so equal applies to >(=)
+        std::function<bool(int, int)> compare;
```

Comment:
std::function introduces overhead, try to avoid it whenever possible. In this case, `(a < b) ^ performancePreferred` will do.